[45] isChainValid method must also check blocks chaining

jhonderson opened the pull request:
`isChainValid` method must also check that for each block in the chain its Block.previousHash actually match the Block.hash of its previous block.

https://github.com/Savjee/SavjeeCoin/issues/45

KaffinPX reviewed (changes requested):
What about removing package-lock :9
and it will LGTM

jhonderson commented:
> What about removing package-lock :9
> and it will LGTM

Done, thanks

KaffinPX commented:
@Savjee Now its legit, can u pull it?

Savjee commented on src/blockchain.js:
```diff
@@ -267,6 +267,11 @@ class Blockchain {
       const currentBlock = this.chain[i];
+      const previousBlock = this.chain[i - 1];
+
+      if (previousBlock.hash !== currentBlock.previousHash) {
```
Could you use the `calculateHash()` method here instead?
Otherwise it would still be possible to manually set a correct hash, even if it doesn't match with the contents of the block.

KaffinPX replied:
Right

jhonderson replied:
Hello @Savjee ,
I think that is already validated at line 280.

For the previous block the hash value was validated in the line 280 in the previous iteration. For the current block, it will validated at line 280 in the current iteration. The first block is the génesis one, I don’t think there is a problem with that.

Adding calculateHash() does not affect too much the time complexity of this validation method, but better not call it if not necessary right?

jhonderson replied:
Maybe one thing we are missing in this method is validate that hash match the configured difficulty, so that if the attacker changes the block information and recalculate all the hashes, they are still  compliant with the difficulty restriction.

If I am correct we may creat a different task/pr for that.

Savjee replied:
You're right, we don't need to recalculate it in the loop. Nice catch!

Yes, the difficulty check is a good idea. And should indeed happen in a separate PR. I'll merge this one in first ;)

Savjee reviewed (commented):
Looks good! Just 1 small remark and then it's ready to be merged.
Could you also merge in the latest master branch? I added some CI based on GitHub Actions.

jhonderson commented:
> Looks good! Just 1 small remark and then it's ready to be merged.
> Could you also merge in the latest master branch? I added some CI based on GitHub Actions.

Updated latest changes from master branch :)

KaffinPX commented:
@Savjee Are u planning to adding P2P using net TCP?
I have a prototype i can give u if u want Discord: KaffinPX#0001

Savjee commented:
Thanks again!

@KaffinPX I have no plans to do that, but I'm open to pull requests. Ideally the P2P aspect should be optional.
